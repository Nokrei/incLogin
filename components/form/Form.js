import { useContext } from "react";
import { useForm } from "react-hook-form";
import AuthContext from "../../context/AuthContext";
import styles from "./Form.module.css";

export default function Form() {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm();

  const { login, error } = useContext(AuthContext);

  const submitForm = (user) => {
    login(user);
  };

  return (
    <form onSubmit={handleSubmit(submitForm)} className={styles.form}>
      <div className={styles.inputGroup}>
        <label htmlFor="username">Username</label>
        <input
          type="text"
          id="username"
          {...register("username", { required: true, maxLength: 20 })}
        />
        {errors.username && (
          <span className={styles.error}>Please input username</span>
        )}
      </div>
      <div className={styles.inputGroup}>
        <label htmlFor="password">Password</label>
        <input
          type="password"
          id="password"
          {...register("password", { required: true, maxLength: 16 })}
        />
        {errors.password && (
          <span className={styles.error}>Please input valid password</span>
        )}
      </div>
      <p className={styles.error}>{error}</p>
      <button type="submit" className="btn btn-primary">
        Log in
      </button>
    </form>
  );
}
