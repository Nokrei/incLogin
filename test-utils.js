// https://testing-library.com/docs/react-native-testing-library/setup/#custom-render
import { render } from "@testing-library/react";
import { AuthProvider } from "./context/AuthContext";

const AllTheProviders = ({ children, customContext }) => {
  return (
    <AuthProvider
      value={{
        user: null,
        error: false,
        login: () => null,
        logout: () => null,
        loading: false,
        ...customContext,
      }}
    >
      {children}
    </AuthProvider>
  );
};

const customRender = (ui, customContext, options) =>
  render(ui, {
    wrapper: () => (
      <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>
    ),
    ...options,
  });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
