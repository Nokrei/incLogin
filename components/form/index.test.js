import { render, screen, waitFor } from "../../test-utils";
import Form from "./Form";

describe("Login form", () => {
  it("should render", () => {
    const { container } = render(<Form />, { loading: true });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should login", async () => {
    const mockLogin = jest.fn();
    render(<Form />, { login: mockLogin });

    userEvent.type(screen.getByPlaceholderText("Email"), "dan@fake.com");
    userEvent.click(screen.getByText("Login"));

    await waitFor(() =>
      expect(mockLogin).toHaveBeenCalledWith({ email: "dan@fake.com" })
    );
  });
});
