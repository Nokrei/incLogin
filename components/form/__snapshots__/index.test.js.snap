// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login form should render 1`] = `
<form
  class="form"
>
  <div
    class="inputGroup"
  >
    <label
      for="username"
    >
      Username
    </label>
    <input
      id="username"
      name="username"
      type="text"
    />
  </div>
  <div
    class="inputGroup"
  >
    <label
      for="password"
    >
      Password
    </label>
    <input
      id="password"
      name="password"
      type="password"
    />
  </div>
  <button
    class="btn btn-primary"
    type="submit"
  >
    Log in
  </button>
</form>
`;
